crates/volta-core/src/hook/mod.rs

```diff
@@ -68,27 +67,103 @@ pub struct ToolHooks<D: Distro> {
     pub phantom: PhantomData<D>,
 }
 
+impl<D: Distro> ToolHooks<D> {
+    /// Creates a merged struct, with "right" having precedence over "left".
+    fn merge(left: Self, right: Self) -> Self {
+        Self {
+            distro: right.distro.or(left.distro),
+            latest: right.latest.or(left.latest),
+            index: right.index.or(left.index),
+            phantom: PhantomData,
+        }
+    }
+}
+
 impl HookConfig {
-    /// Returns the current hooks, loaded from the filesystem.
+    /// Returns the current hooks, which are a merge between the user hooks and
+    /// the project hooks (if any).
     fn current() -> Fallible<Self> {
-        let path = user_hooks_file()?;
-        let src = touch(&path)
-            .and_then(|mut file| file.read_into_string())
-            .with_context(|_| ErrorDetails::ReadHooksError {
-                file: path.to_string_lossy().to_string(),
-            })?;
-        src.parse()
+        let maybe_project_config = Self::for_current_dir()?;
+        let user_config = Self::for_user()?;
+
+        Ok(if let Some(project_config) = maybe_project_config {
+            Self::merge(user_config, project_config)
+        } else {
+            user_config
+        })
+    }
+
+    /// Returns the per-project hooks for the current directory.
+    fn for_current_dir() -> Fallible<Option<Self>> {
+        Self::for_dir(&env::current_dir().with_context(|_| ErrorDetails::CurrentDirError)?)
+    }
+
+    /// Returns the per-project hooks for the specified directory.  If the
+    /// specified directory is not itself a project, its ancestors will be
+    /// searched.
+    fn for_dir(base_dir: &Path) -> Fallible<Option<Self>> {
+        match project_for_dir(&base_dir) {
+            Some(project_dir) => {
+                let path = project_dir.join(".volta").join("hooks.json");
+
+                if !path.is_file() {
+                    return Ok(None);
+                }
+
+                Self::from_file(&path).map(|hooks| Some(hooks))
+            }
+            None => Ok(None),
+        }
     }
-}
 
-impl FromStr for HookConfig {
-    type Err = VoltaError;
+    fn from_file(file_path: &Path) -> Fallible<Self> {
+        let file = touch_with_contents(&file_path, "{}").with_context(|_| {
```

Comment:
I keep going back and forth on whether we should do the empty file detection, as opposed to writing a default JSON object. On the one hand, the default object simplifies the logic here a bit since we don't have to do any reading. This is going to be a breaking change anyway, since people will have to migrate their `hooks.toml` to `hooks.json` format, so there isn't any major concern about porting the old empty file to a new one.

On the other hand, I feel like it would be nice if we supported an empty file (or even just no file instead of using `touch`), since that's a more obvious way of saying "No custom hooks" than a specific empty-object value.

@dherman @chriskrycho @mikrostew Thoughts?

Reply:
I've updated this to no longer attempt to create `$VOLTA_HOME/hooks.json` if it isn't present.  It makes the code simpler and cleaner.